metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+        if (obj < best_obj):
+          best_obj = obj
+          best_w = w
+
+      # TODO:
+      # Maybe allow the usage of mini-batch opt?
+
+      idx = rand_int[iter]
+
+      slack_val = 1 + np.matmul(dist_diff[idx, :], w.T)
+
+      if (slack_val > 0):
+        avg_grad_w = (iter * avg_grad_w + dist_diff[idx, :]) / (iter+1)
+      else:
+        avg_grad_w = iter * avg_grad_w / (iter+1)
+
+      scale_f = -np.sqrt(iter+1) / self.gamma
+
+      # proximal operator with negative trimming equivalent
+      w = scale_f * np.minimum(avg_grad_w + self.beta, 0)
+
+    if(self.verbose):
+      print("max iteration reached.")
+
+    # return L matrix yielded from best weights
+    self.components_ = self._get_components(best_w, basis)
+
+    return self
+
+  def _compute_dist_diff(self, T, X, basis):
+    """
+    Helper function to compute the distance difference of every triplet in the
+    space yielded by the basis set.
+    """
+    # Transformation of data by the basis set
+    XB = np.matmul(X, basis.T)
+
+    lenT = len(T)
+    # get all positive and negative pairs with lowest index first
+    # np.array (2*lenT,2)
+    T_pairs_sorted = np.sort(np.vstack((T[:, [0, 1]], T[:, [0, 2]])),
+                             kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs, indices = np.unique(T_pairs_sorted, return_inverse=True,
+                                   axis=0)
+    # calculate L2 distance acording to bases only for unique pairs
+    dist = np.square(XB[uniqPairs[:, 0], :]-XB[uniqPairs[:, 1], :])
+
+    # return the diference of distances between all positive and negative
+    # pairs
+    return dist[indices[:lenT]]-dist[indices[lenT:]]
+
+  def _get_components(self, w, basis):
+    """
+    get components matrix (L) from computed mahalanobis matrix
+    """
+
+    # get rid of inactive bases
+    active_idx, = w > 0
+    w = w[..., active_idx]
+    basis = basis[active_idx, :]
+
+    K, d = basis.shape
+
+    if(K < d):  # if metric is low-rank
+      return np.sqrt(w.T)*basis  # equivalent to np.diag(np.sqrt(w)).dot(B)
+
+    else:   # if metric is full rank
+      return np.linalg.cholesky(np.matmul(basis.T, w.T*basis)).T
+
+  def _to_index_points(self, triplets, X=None):
+    shape = triplets.shape
+    X, triplets = np.unique(np.vstack(triplets), return_inverse=True, axis=0)
+    triplets = triplets.reshape(shape[:2])
+    return triplets, X
+
+  def _initialize_basis(self, triplets, X):
+    """ TODO: complete function description
+    """
+    n_features = X.shape[1]
+
+    # TODO:
+    # Add other options passed as string
+    authorized_basis = ['triplet_diffs']
+    if isinstance(self.basis, np.ndarray):
+      # TODO: should copy?
+      basis = check_array(self.basis, copy=True)
+      n_basis = basis.shape[0]
+      if basis.shape[1] != n_features:
+        raise ValueError('The input dimensionality ({}) of the given '
+                         'linear transformation `init` must match the '
```

Comment:
wrong message. Should refer to the dimensionality of the provided bases

Reply:
My bad!